prost-derive/src/lib.rs

```diff
@@ -196,8 +212,12 @@ fn try_message(input: TokenStream) -> Result<TokenStream, Error> {
             }
 
             #[inline]
+            #[allow(unused_mut)]
             fn encoded_len(&self) -> usize {
-                0 #(+ #encoded_len)*
+                let mut acc = 0usize #(+ #encoded_len_limited)*;
+                #(acc = acc.checked_add(#encoded_len_unlimited)
+                    .expect("encoded length overflows usize");)*
+                acc
```

Comment:
Yes. For singular numeric fields, this gives an upper bound on their encoded length, so the derive macro can use this information to sum lengths of such fields without overflow checks, as seen above. The lengths of fields for which a limit cannot be known statically, are added with `checked_add` to catch possible overflows.

So if your protobuf message only has fixed-size fields, its `encoded_len` method would not need to use `checked_add` and its performance would not be impacted.

As mentioned in the description to 4901e1a19fbef7d7d2486f8b509d60a717b76481, the whole limit thing is rather an overkill: it would be enough to tell "this field is known to have a small encoding", so lengths of such fields could be summed without checks, and there should never be so many singular fields in a message that this sum would overflow `u32`.

Reply:
> the whole limit thing is rather an overkill

To clarify, I coded that for a more comprehensive approach to determining limits on encoded message lengths, where e.g. if a message type would have a known limit derived from its fields, this could be passed on to determine the limit for another message type containing a field of the first type, and so on. But as mentioned in a comment, this raises safety issues when such information is passed to prost-derive via attributes and may be arbitrarily fudged by the macro user.